404 notifier (#1381)
redpanda-data/documentation: modules/deploy/pages/redpanda/manual/resilience/shadowing-guide.adoc

> = Shadowing Guide

Reviewer: @paulohtb6 Can we talk about where this Resilience section will go? Are you planning to keep it under Deploy and include Shadowing, HA, sizing, etc.? Maybe it makes sense for SM if there's separate info for K8s, but for Cloud I think we want it above this, like what we talked about [here](https://developer.hashicorp.com/consul/docs/concept/reliability).

Reviewer: And I agree with Joyce that this page should be nested under shadowing.adoc. For title, I like **Configure Shadowing for Disaster Recovery**.

Reviewer 2: @paulohtb6 Not sure why you are using "resilience" in this path--there is no resilience guide.
Let's keep the logic of the navtree in tact here. Currently, under both Cloud and Self-Managed > Depoly, we see possible platforms/products from which you can deploy. So placing Shadowing here under Linux breaks our org model. 
Please place the entry for Shadowing (or Disaster Recovery) under Manage:
```
Manage
 --> Shadowing
           --> Failover Quickstart (what you are now calling the Shadowing Guide)
```

# If primary cluster is down, check shadow cluster health
rpk cluster info --brokers <shadow-cluster-brokers>
----

**Decision point**: If the primary cluster is responsive, consider whether failover is actually needed. Partial outages may not require full disaster recovery.

[[verify-shadow-status]]
=== Verify shadow cluster status

Check the health of your shadow links:

[,bash]
----
# List all shadow links
rpk shadow list

# Check the status of your disaster recovery link
rpk shadow status <disaster-recovery-link-name>
----

Verify that the following conditions exist before proceeding with failover:

* Shadow link state should be `ACTIVE`.
* Topics should be in `ACTIVE` state (not `FAULTED`).
* Replication lag should be reasonable for your RPO requirements.

[[document-state]]
=== Document current state

Record the current lag and status before proceeding:

[,bash]
----
# Capture current status for post-mortem analysis
rpk shadow status <disaster-recovery-link-name> > failover-status-$(date +%Y%m%d-%H%M%S).log
----

IMPORTANT: Note the replication lag to estimate potential data loss.

[[initiate-failover]]
=== Initiate failover

> If your data shows that you require a complete cluster failover (recommended during disasters):

Reviewer 3: ```suggestion
A complete cluster failover is appropriate If you observe that the source cluster is no longer reachable:
```

[[verify-functionality]]
=== Verify application functionality

Test critical application workflows:

[,bash]
----
# Verify applications can produce messages
rpk topic produce <critical-topic> --brokers <shadow-cluster-brokers>

# Verify applications can consume messages
rpk topic consume <critical-topic> --brokers <shadow-cluster-brokers> --num 1
----

Test message production and consumption, consumer group functionality, and critical business workflows to ensure everything is working properly.

[[cleanup-stabilize]]
=== Clean up and stabilize

After all applications are running normally:

[,bash]
----
# Optional: Delete the shadow link (no longer needed)
rpk shadow delete <disaster-recovery-link-name> --no-confirm
----

Document the time of failover initiation and completion, applications affected and recovery times, data loss estimates based on replication lag, and issues encountered during failover.

> == Troubleshooting common issues

Reviewer: ```suggestion
== Troubleshoot common issues
```

If topics remain stuck after addressing these cluster health issues and you need immediate failover, you can force delete the shadow link to failover all topics:

[,bash]
----
# Force delete the shadow link to failover all topics
rpk shadow delete <disaster-recovery-link-name> --force
----

[WARNING]
====
Force deleting a shadow link immediately fails over all topics in the link. This action is irreversible and should only be used when topics are stuck and you need immediate access to all replicated data.
====

=== Application connection failures

**Problem**: Applications cannot connect to shadow cluster after failover

**Solution**: Verify shadow cluster broker endpoints are correct and check security group and firewall rules. Confirm authentication credentials are valid for the shadow cluster and test network connectivity from application hosts.

=== Consumer group offset issues

**Problem**: Consumers start from beginning or wrong positions

**Solution**: Verify consumer group offsets were replicated (check your filters) and use `rpk group describe <group-name>` to check offset positions. If necessary, manually reset offsets to appropriate positions.

> == Recovery planning

Reviewer: I'd consider consolidating these last 2 sections into 1 Next steps section

After successful failover:

> 1. **Assess primary cluster**: Determine root cause of the outage

Reviewer 3: ```suggestion
1. **Assess source cluster**: Determine root cause of the outage
```

> 2. **Plan recovery**: Decide whether to restore primary cluster or establish new primary

Reviewer 3: ```suggestion
2. **Plan recovery**: Decide whether to restore source cluster or promote shadow cluster permanently
```

3. **Data synchronization**: Plan how to synchronize any data produced during failover

> 4. **Reverse replication**: Consider setting up shadowing in the opposite direction

Reviewer 3: ```suggestion
4. **Fail forward**: Create a new shadow link with the failed over shadow cluster as source to maintain a DR cluster
```

= Shadowing Guide
:description: Step-by-step emergency guide for failing over Redpanda shadow links during disasters.
:env-linux: true
:page-categories: Management, High Availability, Disaster Recovery, Emergency Response

[NOTE]
====
include::shared:partial$enterprise-license.adoc[]
====

This guide provides step-by-step procedures for emergency failover when your primary Redpanda cluster becomes unavailable. Follow these procedures only during active disasters when immediate failover is required.

[IMPORTANT]
====

> This is an emergency procedure. For planned failover testing or regular shadow link management, see xref:deploy:redpanda/manual/resilience/shadowing.adoc[]. Ensure you have completed the xref:deploy:redpanda/manual/resilience/shadowing.adoc#disaster-readiness-checklist[disaster readiness checklist] before an emergency occurs.

Reviewer 2: ```suggestion
This is an emergency procedure. For planned failover testing or day-to-day shadow link management, see xref:deploy:redpanda/manual/resilience/shadowing.adoc[]. Ensure you have completed the xref:deploy:redpanda/manual/resilience/shadowing.adoc#disaster-readiness-checklist[disaster readiness checklist] before an emergency occurs.
```

====

== Emergency failover procedure

> Follow these steps in order during an active disaster:

Reviewer 2: ```suggestion
Follow these steps during an active disaster:
```

Reviewer 2: The fact that they are numbered steps shows they are sequential.

1. <<assess-situation,Assess the situation>>
2. <<verify-shadow-status,Verify shadow cluster status>>  
3. <<document-state,Document current state>>
4. <<initiate-failover,Initiate failover>>
5. <<monitor-progress,Monitor failover progress>>
6. <<update-applications,Update application configuration>>
7. <<verify-functionality,Verify application functionality>>
8. <<cleanup-stabilize,Clean up and stabilize>>

[[assess-situation]]
=== Assess the situation

Confirm that failover is necessary:

[,bash]
----
# Check if the primary cluster is responding
rpk cluster info --brokers <primary-cluster-brokers>

# If primary cluster is down, check shadow cluster health
rpk cluster info --brokers <shadow-cluster-brokers>
----

**Decision point**: If the primary cluster is responsive, consider whether failover is actually needed. Partial outages may not require full disaster recovery.

[[verify-shadow-status]]
=== Verify shadow cluster status

Check the health of your shadow links:

[,bash]
----
# List all shadow links
rpk shadow list

> # Check status of your disaster recovery link

Reviewer 2: ```suggestion
# Check the status of your disaster recovery link
```

rpk shadow status <disaster-recovery-link-name>
----

> Verify these conditions before proceeding with failover:

Reviewer 2: ```suggestion
Verify that the following conditions exist before proceeding with failover:
```

Reviewer 2: Also, can you show an example of this output?

* Shadow link state should be `ACTIVE`.
* Topics should be in `ACTIVE` state (not `FAULTED`).
* Replication lag should be reasonable for your RPO requirements.

[[document-state]]
=== Document current state

Record the current lag and status before proceeding:

[,bash]
----
# Capture current status for post-mortem analysis
rpk shadow status <disaster-recovery-link-name> > failover-status-$(date +%Y%m%d-%H%M%S).log
----

> IMPORTANT: Note the replication lag to estimate potential data loss.

Reviewer 2: How much lag is acceptable/unacceptable? I know you say it depends on your RPO requirements, but can you offer users a specific example?

Reviewer 3: We should also inform users on how to see the lag. rpk shadow status will show both status and lag. Lag will not be displayed in time, it will be in the difference of messages according the the HWM (high water mark)

[[initiate-failover]]
=== Initiate failover

> For complete cluster failover (recommended during disasters):

Reviewer 2: ```suggestion
If your data shows that you require a complete cluster failover (recommended during disasters):
```

[,bash]
----
# Fail over all topics in the shadow link
rpk shadow failover <disaster-recovery-link-name> --all --no-confirm
----

> For selective topic failover (if only specific services are affected):

Reviewer 2: ```suggestion
For selective topic failover (when only specific services are affected):
```

Track the promotion process:

[,bash]
----
# Monitor status until all topics show PROMOTED
watch -n 5 "rpk shadow status <disaster-recovery-link-name>"
----

**Wait for**: All critical topics to reach `FAILED_OVER` state before proceeding.

[[update-applications]]
=== Update application configuration

Redirect your applications to the shadow cluster by updating connection strings in your applications to point to shadow cluster brokers. If using DNS-based service discovery, update DNS records accordingly. Restart applications to pick up new connection settings and verify connectivity from application hosts to shadow cluster.

[[verify-functionality]]
=== Verify application functionality

Test critical application workflows:

[,bash]
----
# Verify applications can produce messages
rpk topic produce <critical-topic> --brokers <shadow-cluster-brokers>

# Verify applications can consume messages
rpk topic consume <critical-topic> --brokers <shadow-cluster-brokers> --num 1
----

Test message production and consumption, consumer group functionality, and critical business workflows to ensure everything is working properly.

[[cleanup-stabilize]]
=== Clean up and stabilize

> Once applications are running normally:

Reviewer 2: ```suggestion
After all applications are running normally:
```

[,bash]
----
# Optional: Delete the shadow link (no longer needed)
rpk shadow delete <disaster-recovery-link-name> --no-confirm
----

Document the time of failover initiation and completion, applications affected and recovery times, data loss estimates based on replication lag, and issues encountered during failover.

== Troubleshooting common issues

=== Shadow link in PAUSED state

**Problem**: Shadow link shows `PAUSED` instead of `ACTIVE`

> 

Reviewer 4: PAUSED is not implemented yet (its impossible for the link to be in this state for now)

**Solution**:
[,bash]
----
# Resume the shadow link
rpk shadow resume <disaster-recovery-link-name>

# Wait for it to return to ACTIVE state
rpk shadow status <disaster-recovery-link-name>
----

=== Topics stuck in FAILING_OVER state

**Problem**: Topics remain in `FAILING_OVER` state for extended periods

> **Solution**: Check shadow cluster logs for errors and ensure sufficient resources (CPU, memory, disk) are available on the shadow cluster. Verify network connectivity between shadow cluster nodes.

Reviewer 4: also ensure there are no leaderless shadow topic partitions or leaderless /under-replicated controller.

> If topics remain stuck and you need to fail over everything immediately, you can force delete the shadow link, which will promote all topics:

Reviewer 2: ```suggestion
If topics remain stuck and you need to failover everything immediately, you can force delete the shadow link, which will promote all topics:
```

Reviewer 2: [Failover: This is a technical term used in IT and disaster recovery contexts. It describes the process of automatically or manually switching to a standby system, server, or network upon the failure of the primary system. Failover ensures continuity of service and minimizes downtime during outages. For example, if a server goes down, the system can failover to a backup server to maintain operations. ](https://www.bing.com/ck/a?!&&p=4a4e5d7564b5917d1b5eb86e2449e974c6eed361ee9329b42014d884e615835eJmltdHM9MTc2MDY1OTIwMA&ptn=3&ver=2&hsh=4&fclid=1b1804ee-b18d-6bd1-3125-1580b0da6a0f&u=a1aHR0cHM6Ly9sZWFybi5taWNyb3NvZnQuY29tL2VuLXVzL2F6dXJlL3JlbGlhYmlsaXR5L2NvbmNlcHQtZmFpbG92ZXItZmFpbGJhY2s&ntb=1)
[2](https://www.bing.com/ck/a?!&&p=4a4e5d7564b5917d1b5eb86e2449e974c6eed361ee9329b42014d884e615835eJmltdHM9MTc2MDY1OTIwMA&ptn=3&ver=2&hsh=4&fclid=1b1804ee-b18d-6bd1-3125-1580b0da6a0f&u=a1aHR0cHM6Ly9sZWFybi5taWNyb3NvZnQuY29tL2VuLXVzL2F6dXJlL3JlbGlhYmlsaXR5L2NvbmNlcHQtZmFpbG92ZXItZmFpbGJhY2s&ntb=1)
Fail Over: This term is often used informally to describe the action of switching from one system to another. However, it is not as widely recognized or used in formal documentation as "failover." In many cases, "fail over" may be seen as a less precise way to describe the same process.

Reviewer 2: I understand why you put "fail over" in certain contexts throughout, but every time I see it followed by "failover" I find it distracting, or just an inconsistency (which can undermine a reader's confidence). Pretty minor/trivial, but thinking maybe just use failover throughout. At least that is Copilot's guidance. Mine too.

[,bash]
----
# Force delete the shadow link to promote all topics
rpk shadow delete <disaster-recovery-link-name> --force
----

[WARNING]
====
Force deleting a shadow link immediately promotes all topics in the link. This action is irreversible and should only be used when topics are stuck and you need immediate access to all replicated data.
====

=== Application connection failures

**Problem**: Applications cannot connect to shadow cluster after failover

> **Solution**: Verify shadow cluster broker endpoints are correct and check security group and firewall rules. Confirm authentication credentials are valid for shadow cluster and test network connectivity from application hosts.

Reviewer 2: ```suggestion
**Solution**: Verify shadow cluster broker endpoints are correct and check security group and firewall rules. Confirm authentication credentials are valid for the shadow cluster and test network connectivity from application hosts.
```

=== Consumer group offset issues

**Problem**: Consumers start from beginning or wrong positions

> **Solution**: Verify consumer group offsets were replicated (check your filters) and use `rpk group describe <group-name>` to check offset positions. If necessary, manually reset offsets to appropriate positions.

Reviewer 2: Can you add a link to the doc they can follow for manually resetting consumer group offsets?

https://support.redpanda.com/hc/en-us/articles/23499121317399-How-to-manage-consumer-group-offsets-in-Redpanda

> = Shadowing Runbook

Reviewer: @paulohtb6 I think the page should be renamed too, so "emergency" is not in the URL. Also, the term runbook feels internal to me. What do you think about **Failover for Disaster Recovery** or **Disaster Recovery Guide**? cc @Feediver1

Reviewer 5: Changed to "Shadowing Guide". Let me know if it's ok or if I should change more.

* Topics should be in `ACTIVE` state (not `FAULTED`)
* Replication lag should be reasonable for your RPO requirements

=== Step 3: Document current state

Record the current lag and status before proceeding:

[,bash]
----
# Capture current status for post-mortem analysis
rpk shadow status <disaster-recovery-link-name> > failover-status-$(date +%Y%m%d-%H%M%S).log
----

IMPORTANT: Note the replication lag to estimate potential data loss.

=== Step 4: Initiate failover

For complete cluster failover (recommended during disasters):

[,bash]
----
# Fail over all topics in the shadow link
rpk shadow failover <disaster-recovery-link-name> --all --no-confirm
----

For selective topic failover (if only specific services are affected):

[,bash]
----
# Fail over individual topics
rpk shadow failover <disaster-recovery-link-name> --topic <critical-topic-1> --no-confirm
rpk shadow failover <disaster-recovery-link-name> --topic <critical-topic-2> --no-confirm
----

=== Step 5: Monitor failover progress

Track the promotion process:

[,bash]
----
# Monitor status until all topics show PROMOTED
watch -n 5 "rpk shadow status <disaster-recovery-link-name>"
----

> **Wait for**: All critical topics to reach `PROMOTED` state before proceeding.

Reviewer 6: ```suggestion
**Wait for**: All critical topics to reach `FAILED_OVER` state before proceeding.
```

# Verify applications can consume messages
rpk topic consume <critical-topic> --brokers <shadow-cluster-brokers> --num 1
----

**Test**:

* Message production and consumption
* Consumer group functionality
* Critical business workflows

=== Step 8: Clean up and stabilize

After applications are running normally:

[,bash]
----
# Optional: Delete the shadow link (no longer needed)
rpk shadow delete <disaster-recovery-link-name> --no-confirm
----

**Document**:

* Time of failover initiation and completion
* Applications affected and recovery times
* Data loss estimates based on replication lag
* Issues encountered during failover

== Troubleshooting common issues

=== Shadow link in PAUSED state

**Problem**: Shadow link shows `PAUSED` instead of `ACTIVE`

**Solution**:
[,bash]
----
# Resume the shadow link
rpk shadow resume <disaster-recovery-link-name>

# Wait for it to return to ACTIVE state
rpk shadow status <disaster-recovery-link-name>
----

=== Topics stuck in FAILING_OVER state

**Problem**: Topics remain in `FAILING_OVER` state for extended periods

**Solution**:

* Check shadow cluster logs for errors
* Ensure sufficient resources (CPU, memory, disk) on shadow cluster

> * Verify network connectivity between shadow cluster nodes

Reviewer 6: If stuck and they're fine with "failing over everything", then they can force delete the link
